tpls/desul/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,490 @@
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_inc(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_sub(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_dec(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_mul(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr *= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_div(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr /= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_lshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr << val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+DESUL_IMPL_ATOMICS_OPENACC_PREFIX(T,T) device_atomic_fetch_rshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr >> val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_fetch_mod(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  if(acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_fetch_mod(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  tmp = *ptr; *ptr = *ptr % val;
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_fetch_max(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  if(acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_fetch_max(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  tmp = *ptr; *ptr = std::max(*ptr, val);
+  return tmp;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
+#pragma acc routine seq
+inline int device_atomic_fetch_max(
+int* ptr, const int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  int tmp;
+  tmp = atomicMax(ptr, val);
```

Comment:
Presumably this only works with NVHPC

Reply:
Yes, using CUDA atomic intrinsics will work only with NVHPC.